Sources/Towel/DataStructures/AvlTree.cs

```diff
@@ -517,13 +537,33 @@ internal static Node RemoveRightMost(Node node, out Node rightMost)
 		/// <inheritdoc/>
 		public T[] ToArray()
 		{
-			#warning TODO: optimize
-			T[] array = new T[_count];
-			int i = 0;
-			this.Stepper(x => array[i++] = x);
-			return array;
+			if (_root == null)
+			{
+				return Array.Empty<T>();
+			}
+			else
+			{
+				T[] array = new T[_count];
+				int i = 0;
+				StackArray<AvlTreeLinked<T, TCompare>.Node> TreeStack = new(_count);
```

Comment:
This should not need to be `_count`: `StackArray<AvlTreeLinked<T, TCompare>.Node> TreeStack = new(_count);`
The stack should only need to be as big as the tree is tall. Using `_count` is allocating a lot of unnecessary memory.
> The height of an AVL tree is bounded by roughly 1.44 * log2 N, while the height of a red-black tree may be up to 2 * log2 N.
> http://pages.cs.wisc.edu/~ealexand/cs367/NOTES/AVL-Trees/index.html#:~:text=The%20height%20of%20an%20AVL%20tree%20is%20bounded%20by%20roughly,to%202%20*%20log2%20N.

Also, I might want to just use an `AvlTreeLinked<T, TCompare>.Node[]` rather than a `StackArray<AvlTreeLinked<T, TCompare>.Node>` as that would be a _very slight_ optimization to not have to jump through another reference. If `StackArray` had a `struct` version that would not be a problem. I have been working on a design to have `struct` versions of the data structures so they can be used more efficiently by other data structures, but I haven't decided on a final design yet.

Reply:
Understood. I'll work on it

Reply:
Replaced it with a simple `StackLinked<>`